design-proposals/klip-33-key-format.md

```diff
@@ -0,0 +1,299 @@
+
+We know from customers and community members that there are a lot of people that have data with 
+non-`KAFKA` formatted keys. This is the first step to unlocking that data and use-cases.
+
+With support for `AVRO` and `JSON` key formats there are a lot of existing use-cases that suddenly
+no longer require pre-processing, or tricky Connect SMTs configured and there are new use-cases,
+which ksqlDB was previously unsuitable for, as documented in the motivation section, which it can 
+now handle.
+
+## Public APIS
+
+### CREATE properties
+
+The following new properties will be accepted in the `WITH` clause of `CREATE` statements for streams
+and tables.
+
+* `KEY_FORMAT`: sets the key format, works long the same lines as the existing `VALUE_FORMAT`.
+* `FORMAT`: sets both the key and value format with a single property.
+
+The existing `VALUE_FORMAT` property will no longer be required, if the server config provides a 
+default. (See below).
+
+### Server configs
+
+The following new configuration options will be added. These configurations can be set globally, 
+within the application property file, or locally, via the `SET` command.
+
+* `ksql.persistence.default.format.key`: the default key format.
+* `ksql.persistence.default.format.value`: the default value format.
+
+## Design
+
+### New CREATE properties
+
+The new `KEY_FORMAT` or `FORMAT` property will be supported where ever the current `VALUE_FORMAT` is
+supported. Namely:
+
+ * In `CRREATE STREAM` and `CREATE TABLE` statements.
+ * In `CREATE STREAM AS SELECT` and `CREATE TABLE AS SELECT` statements.
+ 
+Providing `FORMAT` will set both the key and value formats. Providing `FORMAT` along with either 
+`KEY_FORMAT` or `VALUE_FORMAT` will result in an error.
+
+The key format will follow the same inheritance rules as the current value format. Namely: any 
+derived stream will inherit the format of its leftmost source, unless the format is explicitly set
+in the with clause.
+
+For example:
+
+```sql
+-- Creates a table over a changelog topic with AVRO key and JSON value:
+CREATE TABLE USERS (
+    ID BIGINT PRIMARY KEY,
+    NAME STRING
+  ) WITH (
+    KAFKA_TOPIC='USERS',
+    KEY_FORMAT='AVRO',
+    VALUE_FORMAT='JSON'
+  );
+
+-- Creates a stream over a topic with JSON key and value:
+CREATE STREAM BIDS (
+    ITEM_ID BIGINT KEY,
+    USER_ID BIGINT,
+    AMMOUNT INT
+  ) WITH (
+    KAFKA_TOPIC='USERS',
+    FORMAT='JSON'
+  );
+
+-- Change the key format of a stream:
+CREATE STREAM AVRP_BIDS 
+  WITH WITH (
+    KEY_FORMAT='AVRO'
+  ) AS
+    SELECT * FROM BIDS;
+
+-- Creates an enriched stream. The key format is inherited from the leftmost source, i.e. JSON:
+CREATE STREAM ENRICHED_BIDS AS 
+  SELECT *
+  FROM BIDS 
+   JOIN USERS ON BIDS.USER_ID = USERS.ID;
+```
+
+For formats that support integration with the schema registry, the key schema will be read and 
+registered with the Schema Registry as needed, following the same pattern as the value schema in 
+the current product.
```

Comment:
This is also an interesting point with regard to _when_ the format changes happens -- internal topics might register different stuff in the SR -- if we let the optimizer decide, and we change the optimization rules in the future, it raises schema compatibility questions.

Reply:
Nope, no concerns at all.   See the compatibility section.

Reply:
Again, see https://github.com/confluentinc/ksql/blob/dcd28ff1b26705625a477db4e3e81a1891b184aa/design-proposals/klip-33-key-format.md#internal-topics